### PR TITLE
MAINT: Minor ufunc cleanup

### PR DESCRIPTION
**Pull request by eric-wieser**
* Pulling a long chunk of code into its own function
* Using the same name everywhere when `ufunc->name == NULL`

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -1974,6 +1981,123 @@ _check_ufunc_fperr(int errmask, PyObject *extobj, const char *ufunc_name) {
...
+ *    In previous versions, core dimensions of size 1 would broadcast
+ *    against other core dimensions with the same label.
+ *  * All core dimensions must have their size specified by a passed in
+ *    input or output argument. In previous versions, core dimensions in
+ *    an output argument that were not specified in an input argument,
+ *    and whose size could not be inferred from a passed in output
+ *    argument, would have their size set to 1.
+ */
+static int
+_get_coredim_sizes(PyUFuncObject *ufunc, PyArrayObject **op,
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Almost all of this function, and its doc-comment, are a direct move of the code from before, with the exception that:
* `goto fail` is just `return -1`
* some fields need re-extracting from ufunc
* `ufunc_name` is recalculated when needed - only in errors, so performance is irrelevant

File: numpy/core/src/umath/ufunc_object.c [hunk 2]
```diff
@@ -1983,7 +2107,7 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
     int nin, nout;
     int i, j, idim, nop;
     const char *ufunc_name;
-    int retval = -1, subok = 1;
+    int retval = 0, subok = 1;
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Starting at 0 makes more sense, because our function works by assigning to it until its not zero

Comment (jaimefrio) [inline numpy/core/src/umath/ufunc_object.c]: It's always overwritten, so the value assigned here is irrelevant, right?

**Review by jaimefrio — APPROVED** [review]
Even after this it's still almost a 500 LOC function, way too much, so this is a step in the right direction, I would say.

**Comment by mhvk** [comment]
I see the point, though #8819 will mean that it will need do_remap and remap_axis arguments, which means it will be less stand-alone than it is now.

**Comment by eric-wieser** [comment]
I don't think that needing to add extra parameters in future is an argument against this function - even if it needs more inputs, it still has one goal, which is to produce the argument passed to the inner loop.

Also, when passed `do_remap` and `remap_axis`, they could be marked as `const int`, which would be clearer about how they are used in which bits of the code

File: numpy/core/src/umath/ufunc_object.c [hunk 3]
```diff
@@ -2089,109 +2213,9 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
...
-            }
-        }
-        PyErr_Format(PyExc_ValueError,
-                     "%s: Output operand %d has core dimension %d "
-                     "unspecified, with gufunc signature %s",
-                     ufunc_name, out_op, i, ufunc->core_signature);
+    /* Collect the lengths of the labelled core dimensions */
+    retval = _get_coredim_sizes(ufunc, op, core_dim_sizes);
+    if(retval < 0) {
         retval = -1;
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: I don't follow - why not?

File: numpy/core/src/umath/ufunc_object.c (hunk 2 above)
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: True, but this the correct base case if we end up with a code path that never sets `retval`, so seems like a good thing to keep

File: numpy/core/src/umath/ufunc_object.c (hunk 3 above)
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Oh, wait, yes I do...

**Comment by eric-wieser** [comment]
@jaimefrio: Fixed that stray `retval = -1` line. Good to merge?

**Comment by eric-wieser** [comment]
@mhvk: Are you against merging this before #8819?

**Comment by mhvk** [comment]
Not at all -- I think that one hasn't converged, and I can rebase if need be. My main point was that, with it, the case for splitting this part off became a bit less strong, since now more of the function's state has to be passed on.

**Comment by eric-wieser** [comment]
> since now more of the function's state has to be passed on.

This was more about making it clear which state these loops are producing, not about what state they take as input

**Comment by eric-wieser** [comment]
I'd like to pull the trigger on this before we accumulate more technical debt in the ufunc machinery

**Comment by mhvk** [comment]
I'm OK with this being merged; and certainly we should merge the name part.

**Comment by eric-wieser** [comment]
In it goes then!

**Pull request merged**